lib/Sema/CSDiag.cpp

```diff
@@ -5754,12 +5754,49 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     return true;
   }
   
+  // Check if any of the candidates has a specified number of arguments.
+  auto hasMatchingArgNum = [&](CalleeCandidateInfo &CCI, size_t numArgs) {
+    for (unsigned i = 0; i < CCI.size(); ++i) {
+      auto &&candidate = CCI[i];
+      if (auto *ty  = candidate.getUncurriedFunctionType()) {
+        if (auto *FTy = ty->getAs<AnyFunctionType>()) {
+          if (numArgs == FTy->getParams().size()) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  };
+
   bool hasTrailingClosure = callArgHasTrailingClosure(callExpr->getArg());
   
   // Collect a full candidate list of callees based on the partially type
   // checked function.
   CalleeCandidateInfo calleeInfo(fnExpr, hasTrailingClosure, CS);
 
+  // In the case that function subexpression was resolved independently in
+  // the first place, the resolved type may not provide the best diagnostic.
+  // We consider the number of arguments to decide whether we'd go with it or
+  // stay with the original one.
+  if (fnExpr != callExpr->getFn()) {
```

Comment:
Just counting arguments isn't necessarily going to work for default arguments and variadics.

Reply:
I agree with @CodaFi here, unfortunately it's not as simple as only matching number of parameters, it needs to account for defaults and variadics as well.

Reply:
As an alternative approach - just add all of the overloads from original function to existing `calleeInfo` and filter via `evaluateCloseness` with the rest?